Polygon::isConvexHull(), convexHull(), orientation(), unit tests

IlliaHryshchuk opened the pull request:
Completed:
 - Polygon::convexHull(); (without openMP, I did not come up with idea how to effectively implement it... any ideas are welcome)
 - Polygon::isConvexHull(); (with openMP)
 - Polygon::orientation(); (with openMP)

Added:
 - Polygon::AddRange(const std::vector& p);

Test cases:
 - PolygonAddRange
 - PolygonOrientation
 - PolygonOrientation_false
 - PolygonConvexHull
 - PolygonIsConvexHull
 - PolygonIsConvexHull_false

IlliaHryshchuk committed:
Polygon::isConvexHull(), convexHull(), orientation(), unit tests

VladGribinchuk commented on GeometryCore/include/Polygon.h:
```diff
@@ -44,6 +44,7 @@ namespace geom_utils
         }
 
         void add(const FPoint2D& p) { points.push_back(p); }
+        void addRange(const std::vector<FPoint2D>& p) { std::for_each(p.begin(), p.end(), [&](auto el) { points.push_back(el); }); }
```
`std::copy(p.begin(), p.end(), std::back_inserter(points));` would be a bit neater, isn't it?)
`copy` is exactly what we want to do here

IlliaHryshchuk replied:
yes, it is better. i will implement it

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -27,8 +27,20 @@ namespace geom_utils
     bool Polygon::orientation() const
-    { 
-        return {}; 
+    {
```
Looks like the `float res` here will be equal to the double area of polygon, if i'm not mistaken. So, this code(with some corrections) **should** be moved to `Polygon::area()` instead. And result of `Polygon::orientation()` method could be easily returned via `area()` call. Since you've started implementing `orientation()`, so please complete the `area()` also. You will ease a bit the work for another dev)

IlliaHryshchuk replied:
is it OK, if the Polygon::area() returns negative result? or maybe it is better to left Polygon::orientation() without changes, and use this code for Polygon::area() + modification (return abs(result/2); ?

VladGribinchuk replied:
It is totally OK: positive/negative area defines polygone orientation ccw/cw(counter clockwise/clockwise). This is widely used in different graphic libs. As a rule, graphic libs have convention: ccw polygons denote external contours and cw polygons denote holes. External contour could have hole(s) inside. Total area of group of polygons is just sum of all polygons' areas, and it works as expected even if there are holes(those will have negative area value).

IlliaHryshchuk replied:
oh, i have understood. I will implement it

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -27,8 +27,20 @@ namespace geom_utils
+            res += (v2.x - v1.x) * (v2.y + v1.y);
+        }
+
+        return res < 0 ? true : false;
```
`return res < 0 ? true : false;` is verbose. `return res < 0;` will have the same effect.

IlliaHryshchuk replied:
oh, it is really better, i will implement it

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -58,13 +70,55 @@ namespace geom_utils
     Polygon Polygon::convexHull() const 
     { 
-        return {}; 
+        //if (points.size() < 3) return; //can't bild a polygon from 2 points or less //idk how to do it
```
Maybe just return `*this`? (if (points.size() < 3)

IlliaHryshchuk replied:
Yes, it works. thank you)

VladGribinchuk replied:
Even for 3 points there is no sense to calculate convex hull. So, in fact, condition should be `if (points.size() <= 3)`

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -58,13 +70,55 @@ namespace geom_utils
+        Polygon convexHull;
+        int pos_of_the_most_right_point = 0;
+
+        for (int i = 1; i < points.size(); i++) if (points[i].x < points[pos_of_the_most_right_point].x) pos_of_the_most_right_point = i; //idk how to do it with stl
```
It is ok.
But you can also use `std::min_element()` which will explicitly state our intention of searching for some minimum element in a range. As a predicate for `std::min_element()` you can use your comparison of `x` component.

IlliaHryshchuk replied:
oh, it looks great. i will implement it

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -58,13 +70,55 @@ namespace geom_utils
+            FPoint2D v1 = points[i];
+            FPoint2D v2 = points[(i + 1) % points.size()];
+            FPoint2D v3 = points[(i + 2) % points.size()];
+            float res = (v2.x - v1.x) * (v3.y - v2.y) - (v2.y - v1.y) * (v3.x - v2.x);
```
This is calculation of cross product of two vectors. It is already present in GeometryCore, please see LinearAlgebra.h and use corresponding function.

IlliaHryshchuk replied:
really, i will use it

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -58,13 +70,55 @@ namespace geom_utils
+
+            #pragma omp critical
+            {
+                if (res < 0) convex_flag |= 1;
```
This section and final check look somewhat overloaded. From my view, `Polygon::isConvexHull()` could be simplified.
Please consider the following pseudocode. 
```
return true if:
    orientation of polygon is counter clockwise and all cross products are non-negative 
        or
    orientation of polygon is clockwise and all cross products are non-positive
otherwise return false
```

IlliaHryshchuk replied:
ok, i will try to do it

VladGribinchuk commented on UnitTests/PolygonTests.h:
```diff
@@ -41,3 +41,45 @@ DEFINE_TEST_CASE(PolygonInput_CorruptedData)
+DEFINE_TEST_CASE(PolygonOrientation_false)
+{
+    Polygon p({ {-2.000, 5.000}, { 3.000, 5.000 }, { 3.000, -1.000 }, { -1.000, -2.000 } });
+    TEST_ASSERT(!(p.orientation()), "Orientation must be non-counter clockwise!");
```
Orientation must be **non-counter clockwise**!
... must be **clockwise**

IlliaHryshchuk replied:
fixed

VladGribinchuk reviewed (changes requested):
Please see some of my comments/requests

IlliaHryshchuk committed:
improoved:
 - Polygon::isConvexHull(), convexHull(), orientation(), addRange();

completed:
 - Polygon::area();

added:
 - PolygonArea unit test

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -22,13 +23,26 @@ namespace geom_utils
     inline FPoint2D::coord Polygon::area() const 
     { 
-        return {}; 
+        if (this->points.size() < 3) return false; //can't bild a polygon from 2 points or less
```
Polygon::area() should return FPoint2D::coord, not boolean.

IlliaHryshchuk replied:
fixed

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -22,13 +23,26 @@ namespace geom_utils
-        return {}; 
+    {
+        auto res = this->area();
+        return res < 0;
```
Please consider one-line statement: `return area() < 0;`

IlliaHryshchuk replied:
fixed

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -58,13 +72,47 @@ namespace geom_utils
-        return {}; 
+        if (points.size() <= 3) return *this; //can't bild a polygon from 2 points or less
+        Polygon convexHull;
+        auto pos_of_the_most_right_point = min_element(points.begin(), points.end(), [](auto a, auto b) {return b.x > a.x; }) - points.begin();
```
I think there is missing `#inlcude <algorithm>` for `min_element`. Also, `min_element` is located inside `std::` namespace.

IlliaHryshchuk replied:
But i did not have any errors. May be an error did not occur because the `<algorithm>` is included in `Point.h`, and `Point.h` is included in `Polygon.h`?
Anyway, i will add `std::` and `<algorithm>`

VladGribinchuk replied:
> But i did not have any errors. May be an error did not occur because the `<algorithm>` is included in `Point.h`, and `Point.h` is included in `Polygon.h`?

Maybe, thats the case. But never rely that smth is included somewhere, so I dont have to include smth. What if we remove Point.h from Polygon.h or make other changes with includings? This change will trigger other sequence of changes. So, always explicitly include things which you need.

IlliaHryshchuk replied:
OK, i will know it

VladGribinchuk commented on GeometryCore/src/Polygon.cpp:
```diff
@@ -58,13 +72,47 @@ namespace geom_utils
+            if ((isCounterClockwise == true && res < 0) || (isCounterClockwise == false && res > 0)) 
+                isConvex = false;
+        }
+        return isConvex;
```
Good... but still please see some of my suggestions:
1. `if (points.size() < 3) return false;` technically, polygon with 2 points is convex hull, because it doen't have any interior angle bigger than 180)
2. I believe, we can remove `omp` directives here and still have the same good performance. This will allow us to perform early exit from the loop, while the omp loop should carry out all iterations (omp doesn't allow returning from parallel loop). 
Please consider my suggestion below:
```
        bool ccw = orientation();
        for (size_t i = 0; i < points.size(); i++)
        {
            auto vec1 = points[(i + 1) % points.size()] - points[i];
            auto vec2 = points[(i + 2) % points.size()] - points[i];
            if (ccw ? cross(vec1, vec2) < 0 : cross(vec1, vec2) > 0)
                return false;
        }
        return true;
```

IlliaHryshchuk replied:
i was trying to do it with openMP, so i had to add `isConvex` field, and perform all the iterations.
Ok, i will implement it

VladGribinchuk replied:
> i was trying to do it with openMP, so i had to add `isConvex` field, and perform all the iterations.

Yes, thats understandable, and you are right) I just suggested approach without using opemp, which shouldn't loose in performance)

IlliaHryshchuk replied:
Yes, in this case even code has better look)

VladGribinchuk reviewed (commented):
In general, now is better. But please see my comments

IlliaHryshchuk committed:
code optimization and fixes

Polygon::area() : fixed returning value;
Polygon::orientation() : code optimization;
Polygon::convexHull() : fixed namespace for min_element();
Polygon::isConvexHull() : code optimization;

VladGribinchuk commented:
wont wait for other dev reviews to speed up a bit our development process.
I've checked changes and those look good, so I'll merge PR
